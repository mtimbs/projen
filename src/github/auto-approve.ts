import { Component } from '../component';
import { GitHub } from './github';
import { Job, JobPermission } from './workflows-model';

/**
 * Options for 'AutoApprove'
 */
export interface AutoApproveOptions {
  /**
   * Only pull requests authored by these Github usernames will be auto-approved.
   * @default ['github-bot']
   */
  readonly allowedUsernames?: string[];

  /**
   * Only pull requests with this label will be auto-approved.
   * @default 'auto-approve'
   */
  readonly label?: string;

  /**
   * A GitHub secret name which contains a GitHub Access Token
   * with write permissions for the `pull_request` scope.
   *
   * This token is used to approve pull requests.
   *
   * Github forbids an identity to approve its own pull request.
   * If your project produces automated pull requests using the Github default token -
   * {@link https://docs.github.com/en/actions/reference/authentication-in-a-workflow `GITHUB_TOKEN` }
   * - that you would like auto approved, such as when using the `depsUpgrade` property in
   * `NodeProjectOptions`, then you must use a different token here.
   */
  readonly secret: string;
}

/**
 * Auto approve pull requests that meet a criteria
 */
export class AutoApprove extends Component {
  public readonly label: string;

  constructor(github: GitHub, options: AutoApproveOptions) {
    super(github.project);

    this.label = options.label ?? 'auto-approve';
    const usernames = options.allowedUsernames ?? ['github-actions[bot]'];

    let condition = `contains(github.event.pull_request.labels.*.name, '${this.label}')`;
    if (usernames.length > 0) {
      condition += ' && (';
      condition += usernames.map(u => `github.event.pull_request.user.login == '${u}'`).join(' || ');
      condition += ')';
    }

    const approveJob: Job = {
      runsOn: 'ubuntu-latest',
      permissions: {
        pullRequests: JobPermission.WRITE,
      },
      if: condition,
      steps: [{
        uses: 'hmarr/auto-approve-action@v2.1.0',
        with: {
          'github-token': `\${{ secrets.${options.secret} }}`,
        },
      }],
    };

    const workflow = github.addWorkflow('auto-approve');
    workflow.on({
      // The 'pull request' event gives the workflow 'read-only' permissions on some
      // pull requests (such as the ones from dependabot) when using the `GITHUB_TOKEN`
      // security token. This prevents the workflow from approving these pull requests.
      // Github has placed this guard so as to prevent security attacks by simply opening
      // a pull request and triggering a workflow on a commit that was not vetted to make
      // unintended changes to the repository.
      //
      // Instead use the 'pull request target' event here that gives the Github workflow
      // 'read-write' permissions. This is safe because, this event, unlike the 'pull request'
      // event references the BASE commit of the pull request and not the HEAD commit.
      pullRequestTarget: {
        types: ['labeled', 'opened', 'synchronize', 'reopened', 'ready_for_review'],
      },
    });
    workflow.addJobs({ approve: approveJob });
  }
}
