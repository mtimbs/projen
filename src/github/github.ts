import { Component } from '../component';
import { Project } from '../project';
import { Dependabot, DependabotOptions } from './dependabot';
import { Mergify } from './mergify';
import { PullRequestTemplate } from './pr-template';
import { GithubWorkflow } from './workflows';

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;

  /**
   * Enables GitHub workflows. If this is set to `false`, workflows will not be created.
   *
   * @default true
   */
  readonly workflows?: boolean;
}

export class GitHub extends Component {
  /**
   * The `Mergify` configured on this repository. This is `undefined` if Mergify
   * was not enabled when creating the repository.
   */
  public readonly mergify?: Mergify;

  /**
   * Are workflows enabled?
   */
  public readonly workflows: boolean;

  public constructor(project: Project, options: GitHubOptions = {}) {
    super(project);

    this.workflows = options.workflows ?? true;

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this);
    }
  }

  public addWorkflow(name: string) {
    return new GithubWorkflow(this, name);
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }
}
