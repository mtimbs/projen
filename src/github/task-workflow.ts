import { Task } from '../tasks';
import { GitHub } from './github';
import { GithubWorkflow } from './workflows';
import { ContainerOptions, Job, JobPermissions, JobStep, JobStepOutput, Triggers } from './workflows-model';

const DEFAULT_JOB_ID = 'build';
const UBUNTU_LATEST = 'ubuntu-latest';

export interface TaskWorkflowOptions {
  /**
   * The workflow name.
   */
  readonly name: string;

  /**
   * The primary job id.
   * @default "build"
   */
  readonly jobId?: string;

  /**
   * @default - default image
   */
  readonly container?: ContainerOptions;

  /**
   * Adds an 'if' condition to the workflow.
   */
  readonly condition?: string;

  /**
   * A directory name which contains artifacts to be uploaded (e.g. `dist`).
   * If this is set, the contents of this directory will be uploaded as an
   * artifact at the end of the workflow run, even if other steps fail.
   *
   * @default - not set
   */
  readonly artifactsDirectory?: string;

  /**
   * The triggers for the workflow.
   *
   * @default - by default workflows can only be triggered by manually.
   */
  readonly triggers?: Triggers;

  /**
   * Initial steps to run before the source code checkout.
   *
   * @default - not set
   */
  readonly preCheckoutSteps?: JobStep[];

  /**
   * Override for the `with` property of the source code checkout step.
   *
   * @default - not set
   */
  readonly checkoutWith?: Record<string, any>;

  /**
   * Steps to run before the main build step.
   *
   * @default - not set
   */
  readonly preBuildSteps?: JobStep[];

  /**
   * The main task to be executed.
   */
  readonly task: Task;

  /**
   * Actions to run after the main build step.
   *
   * @default - not set
   */
  readonly postBuildSteps?: JobStep[];

  /**
   * Workflow environment variables.
   * @default {}
   */
  readonly env?: Record<string, string>;

  /**
   * Permissions for the build job.
   */
  readonly permissions: JobPermissions;

  /**
   * Mapping of job output names to values/expressions.
   *
   * @default {}
   */
  readonly outputs?: { [name: string]: JobStepOutput };
}

/**
 * A GitHub workflow for common build tasks within a project.
 */
export class TaskWorkflow extends GithubWorkflow {
  private readonly github: GitHub;
  public readonly jobId: string;

  constructor(github: GitHub, options: TaskWorkflowOptions) {
    super(github, options.name);
    this.jobId = options.jobId ?? DEFAULT_JOB_ID;
    this.github = github;

    if (options.triggers) {
      if (options.triggers.issueComment) {
        // https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#potential-impact-of-a-compromised-runner
        throw new Error('Trigger "issueComment" should not be used due to a security concern');
      }

      this.on(options.triggers);
    }

    this.on({
      workflowDispatch: {}, // allow manual triggering
    });

    const preCheckoutSteps = options.preCheckoutSteps ?? [];
    const checkoutWith = options.checkoutWith ? { with: options.checkoutWith } : {};
    const preBuildSteps = options.preBuildSteps ?? [];
    const postBuildSteps = options.postBuildSteps ?? [];

    if (options.artifactsDirectory) {
      postBuildSteps.push({
        name: 'Upload artifact',
        uses: 'actions/upload-artifact@v2.1.1',
        // Setting to always will ensure that this step will run even if
        // the previous ones have failed (e.g. coverage report, internal logs, etc)
        if: 'always()',
        with: {
          name: options.artifactsDirectory,
          path: options.artifactsDirectory,
        },
      });
    }

    const job: Job = {
      runsOn: UBUNTU_LATEST,
      container: options.container,
      env: options.env,
      permissions: options.permissions,
      if: options.condition,
      outputs: options.outputs,
      steps: [
        ...preCheckoutSteps,

        // check out sources.
        {
          name: 'Checkout',
          uses: 'actions/checkout@v2',
          ...checkoutWith,
        },

        // sets git identity so we can push later
        {
          name: 'Set git identity',
          run: [
            'git config user.name "Automation"',
            'git config user.email "github-actions@github.com"',
          ].join('\n'),
        },

        ...preBuildSteps,

        // run the main build task
        {
          name: options.task.name,
          run: this.github.project.runTaskCommand(options.task),
        },

        ...postBuildSteps,
      ],
    };

    this.addJobs({ [this.jobId]: job });
  }
}
