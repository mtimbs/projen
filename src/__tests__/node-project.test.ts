import * as yaml from 'yaml';
import { NodeProject, NodeProjectOptions, LogLevel } from '..';
import { DependencyType } from '../deps';
import { JobPermission } from '../github/workflows-model';
import * as logging from '../logging';
import { NodePackage, NodePackageManager, NpmAccess } from '../node-package';
import { DependenciesUpgradeMechanism } from '../node-project';
import { Project } from '../project';
import { Tasks } from '../tasks';
import { mkdtemp, synthSnapshot, TestProject } from './util';

logging.disable();

test('license file is added by default', () => {
  // WHEN
  const project = new TestNodeProject();

  // THEN
  expect(synthSnapshot(project).LICENSE).toContain('Apache License');
});

test('license file is not added if licensed is false', () => {
  // WHEN
  const project = new TestNodeProject({
    licensed: false,
  });

  // THEN
  const snapshot = synthSnapshot(project);
  expect(snapshot.LICENSE).toBeUndefined();
  expect(snapshot['.gitignore']).not.toContain('LICENSE');
  expect(snapshot['package.json'].license).toEqual('UNLICENSED');
});

describe('deps', () => {

  test('runtime deps', () => {
    // GIVEN
    const project = new TestNodeProject({
      deps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.RUNTIME);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.dependencies).toStrictEqual({
      aaa: '^1.2.3',
      bbb: '~4.5.6',
      ccc: '*',
      ddd: '*',
    });
    expect(pkgjson.peerDependencies).toStrictEqual({});
  });

  test('dev dependencies', () => {
    // GIVEN
    const project = new TestNodeProject({
      devDeps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addDevDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.TEST);
    project.deps.addDependency('eee@^1', DependencyType.DEVENV);
    project.deps.addDependency('fff@^2', DependencyType.BUILD);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.devDependencies.aaa).toStrictEqual('^1.2.3');
    expect(pkgjson.devDependencies.bbb).toStrictEqual('~4.5.6');
    expect(pkgjson.devDependencies.ccc).toStrictEqual('*');
    expect(pkgjson.devDependencies.ddd).toStrictEqual('*');
    expect(pkgjson.devDependencies.eee).toStrictEqual('^1');
    expect(pkgjson.devDependencies.fff).toStrictEqual('^2');
    expect(pkgjson.peerDependencies).toStrictEqual({});
    expect(pkgjson.dependencieds).toBeUndefined();
  });

  test('peerDependencies', () => {
    // GIVEN
    const project = new TestNodeProject({
      peerDeps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addPeerDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.PEER);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.peerDependencies).toStrictEqual({
      aaa: '^1.2.3',
      bbb: '~4.5.6',
      ccc: '*',
      ddd: '*',
    });

    // devDependencies are added with pinned versions
    expect(pkgjson.devDependencies.aaa).toStrictEqual('1.2.3');
    expect(pkgjson.devDependencies.bbb).toStrictEqual('4.5.6');
    expect(pkgjson.devDependencies.ccc).toStrictEqual('*');
    expect(pkgjson.devDependencies.ddd).toStrictEqual('*');
    expect(pkgjson.dependencieds).toBeUndefined();
  });

  test('peerDependencies without pinnedDevDep', () => {
    // GIVEN
    const project = new TestNodeProject({
      peerDependencyOptions: {
        pinnedDevDependency: false,
      },
      peerDeps: [
        'aaa@^1.2.3',
        'bbb@~4.5.6',
      ],
    });

    // WHEN
    project.addPeerDeps('ccc');
    project.deps.addDependency('ddd', DependencyType.PEER);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.peerDependencies).toStrictEqual({
      aaa: '^1.2.3',
      bbb: '~4.5.6',
      ccc: '*',
      ddd: '*',
    });

    // sanitize
    ['npm-check-updates', 'jest', 'jest-junit', 'projen', 'standard-version'].forEach(d => delete pkgjson.devDependencies[d]);

    expect(pkgjson.devDependencies).toStrictEqual({});
    expect(pkgjson.dependencieds).toBeUndefined();
  });

  test('bundled deps are automatically added as normal deps', () => {
    // GIVEN
    const project = new TestNodeProject({
      bundledDeps: ['hey@2.1.1'],
    });

    // WHEN
    project.addBundledDeps('foo@^1.2.3');
    project.deps.addDependency('bar@~1.0.0', DependencyType.BUNDLED);

    // THEN
    const pkgjson = packageJson(project);
    expect(pkgjson.dependencies).toStrictEqual({
      hey: '2.1.1',
      foo: '^1.2.3',
      bar: '~1.0.0',
    });
    expect(pkgjson.bundledDependencies).toStrictEqual([
      'bar',
      'foo',
      'hey',
    ]);
  });
});

test('throw when \'autoApproveProjenUpgrades\' is used with \'projenUpgradeAutoMerge\'', () => {
  expect(() => {
    new TestNodeProject({ autoApproveProjenUpgrades: true, projenUpgradeAutoMerge: true });
  }).toThrow("'projenUpgradeAutoMerge' cannot be configured together with 'autoApproveProjenUpgrades'");
});

describe('deps upgrade', () => {

  test('throws when trying to auto approve projen but auto approve is not defined', () => {
    const message = 'Autoamtic approval of projen upgrades requires configuring `autoApproveOptions`';
    expect(() => { new TestNodeProject({ autoApproveProjenUpgrades: true }); }).toThrow(message);
    expect(() => { new TestNodeProject({ projenUpgradeAutoMerge: true }); }).toThrow(message);
  });

  test('throws when trying to auto approve deps but auto approve is not defined', () => {
    expect(() => {
      new TestNodeProject({ autoApproveUpgrades: true });
    }).toThrow('Autoamtic approval of dependencies upgrades requires configuring `autoApproveOptions`');
  });

  test('workflow can be auto approved', () => {
    const project = new TestNodeProject({
      autoApproveOptions: {
        allowedUsernames: ['dummy'],
        secret: 'dummy',
      },
      autoApproveUpgrades: true,
    });

    const snapshot = yaml.parse(synthSnapshot(project)['.github/workflows/upgrade-dependencies.yml']);
    expect(snapshot.jobs.pr.steps[3].with.labels).toStrictEqual(project.autoApprove?.label);
  });

  test('dependabot can be auto approved', () => {
    const project = new TestNodeProject({
      dependabot: true,
      autoApproveOptions: {
        allowedUsernames: ['dummy'],
        secret: 'dummy',
      },
      autoApproveUpgrades: true,
    });

    const snapshot = yaml.parse(synthSnapshot(project)['.github/dependabot.yml']);
    expect(snapshot.updates[0].labels).toStrictEqual(['auto-approve']);
  });

  test('default - with projen secret', () => {
    const project = new TestNodeProject({ projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN' });
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/workflows/upgrade-dependencies.yml']).toBeDefined();
    expect(snapshot['.github/workflows/upgrade-projen.yml']).toBeUndefined();

    // make sure yarn upgrade all deps, including projen.
    const tasks = snapshot[Tasks.MANIFEST_FILE].tasks;
    expect(tasks['upgrade-dependencies'].steps[2].exec).toStrictEqual('yarn upgrade');
  });

  test('default - no projen secret', () => {
    const project = new TestNodeProject();
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/workflows/upgrade-dependencies.yml']).toBeDefined();
    expect(snapshot['.github/workflows/upgrade-projen.yml']).toBeUndefined();
  });

  test('dependabot - with projen secret', () => {
    const project = new TestNodeProject({
      depsUpgrade: DependenciesUpgradeMechanism.dependabot(),
      projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
    });
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/dependabot.yml']).toBeDefined();
    expect(snapshot['.github/workflows/upgrade-projen.yml']).toBeDefined();
  });

  test('dependabot - no projen secret', () => {
    const project = new TestNodeProject({
      depsUpgrade: DependenciesUpgradeMechanism.dependabot(),
    });
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/dependabot.yml']).toBeDefined();
    expect(snapshot['.github/workflows/upgrade-projen.yml']).toBeUndefined();
  });

  test('github actions - with projen secret', () => {
    const project = new TestNodeProject({
      depsUpgrade: DependenciesUpgradeMechanism.githubWorkflow(),
      projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
    });
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/workflows/upgrade-dependencies.yml']).toBeDefined();
    expect(snapshot['.github/workflows/upgrade-projen.yml']).toBeDefined();
  });

  test('github actions - no projen secret', () => {
    const project = new TestNodeProject({
      depsUpgrade: DependenciesUpgradeMechanism.githubWorkflow(),
    });
    const snapshot = synthSnapshot(project);
    expect(snapshot['.github/workflows/upgrade-dependencies.yml']).toBeDefined();
    expect(snapshot['.github/workflows/upgrade-projen.yml']).toBeUndefined();
  });

  test('throws when depracated dependabot is configued with dependenciesUpgrade', () => {
    expect(() => {
      new TestNodeProject({ dependabot: true, depsUpgrade: DependenciesUpgradeMechanism.githubWorkflow() });
    }).toThrow("'dependabot' cannot be configured together with 'depsUpgrade'");
  });

});

describe('npm publishing options', () => {
  test('defaults', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'my-package',
    });

    // THEN
    expect(npm.npmAccess).toStrictEqual(NpmAccess.PUBLIC);
    expect(npm.npmDistTag).toStrictEqual('latest');
    expect(npm.npmRegistry).toStrictEqual('registry.npmjs.org');
    expect(npm.npmRegistryUrl).toStrictEqual('https://registry.npmjs.org/');
    expect(npm.npmTokenSecret).toStrictEqual('NPM_TOKEN');

    // since these are all defaults, publishConfig is not defined.
    expect(synthSnapshot(project)['package.json'].publishConfig).toBeUndefined();
  });

  test('scoped packages default to RESTRICTED access', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'scoped@my-package',
    });

    // THEN
    expect(npm.npmAccess).toStrictEqual(NpmAccess.RESTRICTED);

    // since these are all defaults, publishConfig is not defined.
    expect(packageJson(project).publishConfig).toBeUndefined();
  });

  test('non-scoped package cannot be RESTRICTED', () => {
    // GIVEN
    const project = new TestProject();

    // THEN
    expect(() => new NodePackage(project, {
      packageName: 'my-package',
      npmAccess: NpmAccess.RESTRICTED,
    })).toThrow(/"npmAccess" cannot be RESTRICTED for non-scoped npm package/);
  });

  test('custom settings', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'scoped@my-package',
      npmDistTag: 'next',
      npmRegistryUrl: 'https://foo.bar',
      npmAccess: NpmAccess.PUBLIC,
      npmTokenSecret: 'GITHUB_TOKEN',
    });

    // THEN
    expect(npm.npmDistTag).toStrictEqual('next');
    expect(npm.npmRegistry).toStrictEqual('foo.bar');
    expect(npm.npmRegistryUrl).toStrictEqual('https://foo.bar/');
    expect(npm.npmAccess).toStrictEqual(NpmAccess.PUBLIC);
    expect(npm.npmTokenSecret).toStrictEqual('GITHUB_TOKEN');
    expect(packageJson(project).publishConfig).toStrictEqual({
      access: 'public',
      registry: 'https://foo.bar/',
      tag: 'next',
    });
  });

  test('deprecated npmRegistry can be used instead of npmRegistryUrl and then https:// is assumed', () => {
    // GIVEN
    const project = new TestProject();

    // WHEN
    const npm = new NodePackage(project, {
      packageName: 'scoped@my-package',
      npmRegistry: 'foo.bar.com',
    });

    // THEN
    expect(npm.npmRegistry).toStrictEqual('foo.bar.com');
    expect(npm.npmRegistryUrl).toStrictEqual('https://foo.bar.com/');
    expect(packageJson(project).publishConfig).toStrictEqual({
      registry: 'https://foo.bar.com/',
    });
  });
});

test('extend github release workflow', () => {
  const project = new TestNodeProject();

  project.release?.addJobs({
    publish_docker_hub: {
      permissions: {
        contents: JobPermission.READ,
      },
      runsOn: 'ubuntu-latest',
      env: {
        CI: 'true',
      },
      steps: [
        {
          name: 'Check out the repo',
          uses: 'actions/checkout@v2',
        },
        {
          name: 'Push to Docker Hub',
          uses: 'docker/build-push-action@v1',
          with: {
            username: '${{ secrets.DOCKER_USERNAME }}',
            password: '${{ secrets.DOCKER_PASSWORD }}',
            repository: 'projen/projen-docker',
            tag_with_ref: 'true',
          },
        },
      ],
    },
  });

  const workflow = synthSnapshot(project)['.github/workflows/release.yml'];
  expect(workflow).toContain('publish_docker_hub:\n    runs-on: ubuntu-latest\n');
  expect(workflow).toContain('username: ${{ secrets.DOCKER_USERNAME }}\n          password: ${{ secrets.DOCKER_PASSWORD }}');
});

describe('scripts', () => {
  test('addTask and setScript', () => {
    const p = new TestNodeProject();
    p.addTask('chortle', { exec: 'echo "frabjous day!"' });
    p.setScript('slithy-toves', 'gyre && gimble');
    const pkg = packageJson(p);
    expect(pkg.scripts).toHaveProperty('chortle');
    expect(pkg.scripts).toHaveProperty('slithy-toves');
  });

  test('removeScript will remove tasks and scripts', () => {
    const p = new TestNodeProject();

    p.addTask('chortle', { exec: 'echo "frabjous day!"' });
    p.setScript('slithy-toves', 'gyre && gimble');
    p.removeScript('chortle');
    p.removeScript('slithy-toves');
    const pkg = packageJson(p);
    expect(pkg.scripts).not.toHaveProperty('chortle');
    expect(pkg.scripts).not.toHaveProperty('slithy-toves');
  });
});

test('mutableBuild will push changes to PR branches', () => {
  // WHEN
  const project = new TestNodeProject({
    mutableBuild: true,
  });

  // THEN
  const workflowYaml = synthSnapshot(project)['.github/workflows/build.yml'];
  const workflow = yaml.parse(workflowYaml);
  expect(workflow.jobs.build.steps).toMatchSnapshot();
});

test('projenDuringBuild can be used to disable "projen synth" during build', () => {
  const enabled = new TestNodeProject({
    projenDuringBuild: true,
  });

  const disabled = new TestNodeProject({
    projenDuringBuild: false,
  });

  const buildTaskEnabled = synthSnapshot(enabled)['.projen/tasks.json'].tasks.build;
  const buildTaskDisabled = synthSnapshot(disabled)['.projen/tasks.json'].tasks.build;
  expect(buildTaskEnabled.steps[0].exec).toEqual('npx projen');
  expect(buildTaskDisabled.steps).toBeUndefined();
});

test('projen synth is only executed for subprojects', () => {
  // GIVEN
  const root = new TestNodeProject();

  // WHEN
  new TestNodeProject({ parent: root, outdir: 'child' });

  // THEN
  const snapshot = synthSnapshot(root);
  const rootBuildTask = snapshot['.projen/tasks.json'].tasks.build;
  const childBuildTask = snapshot['child/.projen/tasks.json'].tasks.build;
  expect(rootBuildTask).toStrictEqual({
    description: 'Full release build (test+compile)',
    name: 'build',
    steps: [{ exec: 'npx projen' }],
  });
  expect(childBuildTask).toStrictEqual({
    description: 'Full release build (test+compile)',
    name: 'build',
  });
});

test('enable anti-tamper', () => {
  // WHEN
  const project = new TestNodeProject({
    packageManager: NodePackageManager.NPM,
    releaseToNpm: true,
    mutableBuild: false,
    antitamper: true,
  });

  // THEN
  const workflowYaml = synthSnapshot(project)['.github/workflows/build.yml'];
  const workflow = yaml.parse(workflowYaml);
  expect(workflow.jobs.build.steps).toMatchSnapshot();
  expect(workflow.jobs.build.steps).toEqual(expect.arrayContaining([
    expect.objectContaining({
      name: 'Anti-tamper check',
    }),
  ]));
});

test('enabling dependabot does not overturn mergify: false', () => {
  // WHEN
  const project = new TestNodeProject({
    dependabot: true,
    mergify: false,
  });

  // THEN
  const snapshot = synthSnapshot(project);
  // Note: brackets important, they prevent "." in filenames to be interpreted
  //       as JSON object path delimiters.
  expect(snapshot).not.toHaveProperty(['.mergify.yml']);
  expect(snapshot).toHaveProperty(['.github/dependabot.yml']);
});

test('github: false disables github integration', () => {
  // WHEN
  const project = new TestNodeProject({
    github: false,
  });

  // THEN
  const output = synthSnapshot(project);
  expect(Object.keys(output).filter(p => p.startsWith('.github/'))).toStrictEqual([]);
});

function packageJson(project: Project) {
  return synthSnapshot(project)['package.json'];
}

class TestNodeProject extends NodeProject {
  constructor(options: Partial<NodeProjectOptions> = {}) {
    super({
      outdir: mkdtemp(),
      name: 'test-node-project',
      logging: {
        level: LogLevel.OFF,
      },
      defaultReleaseBranch: 'master',
      ...options,
    });
  }
}
