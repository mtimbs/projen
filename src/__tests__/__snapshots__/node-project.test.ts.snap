// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enable anti-tamper 1`] = `
Array [
  Object {
    "name": "Checkout",
    "uses": "actions/checkout@v2",
  },
  Object {
    "name": "Set git identity",
    "run": "git config user.name \\"Automation\\"
git config user.email \\"github-actions@github.com\\"",
  },
  Object {
    "name": "Anti-tamper check",
    "run": "git diff --ignore-space-at-eol --exit-code",
  },
  Object {
    "name": "Install dependencies",
    "run": "npm ci",
  },
  Object {
    "name": "build",
    "run": "npx projen build",
  },
  Object {
    "id": "git_diff",
    "name": "Check for changes",
    "run": "git diff --exit-code || echo \\"::set-output name=has_changes::true\\"",
  },
  Object {
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Commit and push changes (if changed)",
    "run": "git add . && git commit -m \\"chore: self mutation\\" && git push origin HEAD:\${{ github.event.pull_request.head.ref }}",
  },
  Object {
    "env": Object {
      "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
    },
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Update status check (if changed)",
    "run": "gh api -X POST /repos/\${{ github.event.pull_request.head.repo.full_name }}/check-runs -F name=\\"build\\" -F head_sha=\\"$(git rev-parse HEAD)\\" -F status=\\"completed\\" -F conclusion=\\"success\\"",
  },
  Object {
    "env": Object {
      "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
    },
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Cancel workflow (if changed)",
    "run": "gh api -X POST /repos/\${{ github.event.pull_request.head.repo.full_name }}/actions/runs/\${{ github.run_id }}/cancel",
  },
  Object {
    "name": "Anti-tamper check",
    "run": "git diff --ignore-space-at-eol --exit-code",
  },
]
`;

exports[`mutableBuild will push changes to PR branches 1`] = `
Array [
  Object {
    "name": "Checkout",
    "uses": "actions/checkout@v2",
    "with": Object {
      "ref": "\${{ github.event.pull_request.head.ref }}",
      "repository": "\${{ github.event.pull_request.head.repo.full_name }}",
    },
  },
  Object {
    "name": "Set git identity",
    "run": "git config user.name \\"Automation\\"
git config user.email \\"github-actions@github.com\\"",
  },
  Object {
    "name": "Install dependencies",
    "run": "yarn install --check-files --frozen-lockfile",
  },
  Object {
    "name": "build",
    "run": "npx projen build",
  },
  Object {
    "id": "git_diff",
    "name": "Check for changes",
    "run": "git diff --exit-code || echo \\"::set-output name=has_changes::true\\"",
  },
  Object {
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Commit and push changes (if changed)",
    "run": "git add . && git commit -m \\"chore: self mutation\\" && git push origin HEAD:\${{ github.event.pull_request.head.ref }}",
  },
  Object {
    "env": Object {
      "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
    },
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Update status check (if changed)",
    "run": "gh api -X POST /repos/\${{ github.event.pull_request.head.repo.full_name }}/check-runs -F name=\\"build\\" -F head_sha=\\"$(git rev-parse HEAD)\\" -F status=\\"completed\\" -F conclusion=\\"success\\"",
  },
  Object {
    "env": Object {
      "GITHUB_TOKEN": "\${{ secrets.GITHUB_TOKEN }}",
    },
    "if": "steps.git_diff.outputs.has_changes",
    "name": "Cancel workflow (if changed)",
    "run": "gh api -X POST /repos/\${{ github.event.pull_request.head.repo.full_name }}/actions/runs/\${{ github.run_id }}/cancel",
  },
]
`;
